### PR TITLE
Template polyfill

### PR DESCRIPTION
**Pull request by dvoytenko**
/cc @erwinmombay

### CODE REVIEW

**Comment by cramforce** [comment]
LGTM

**Comment by erwinmombay** [comment]
same, LGTM

File: src/custom-element.js [hunk 1]
```diff
@@ -332,12 +350,15 @@ export function createAmpElementProto(win, name, implementationClass) {
...
+    }
     const registeredStub = this.implementation_;
     const newImpl = new newImplClass(this);
     this.implementation_ = newImpl;
-    if (registeredStub) {
-      registeredStub.upgrade(newImpl);
-    }
+    this.classList.remove('amp-unresolved');
+    this.classList.remove('-amp-unresolved');
+    newImpl.createdCallback();
```
Comment (erwinmombay) [inline src/custom-element.js]: can we call `this.implementation_.createdCallback()` instead for consistency

File: test/functional/test-custom-element.js [hunk 2]
```diff
@@ -230,6 +242,24 @@ describe('CustomElement', () => {
     expect(testElementBuildCallback.callCount).to.equal(1);
   });
 
+  it('Element - build NOT allowed when in template', () => {
+    const element = new ElementClass();
+    expect(element.classList.contains('-amp-element')).to.equal(true);
+    expect(element.isBuilt()).to.equal(false);
+    expect(element.classList.contains('-amp-notbuilt')).to.equal(true);
+    expect(element.classList.contains('amp-notbuilt')).to.equal(true);
```
Comment (erwinmombay) [inline test/functional/test-custom-element.js]: use `expect(element).to.hav.class('-amp-notbuilt')` instead (here and any other class existence check)

**Comment by erwinmombay** [comment]
minor nits, non merge blocking.

File: src/custom-element.js (hunk 1 above)
Comment (dvoytenko) [inline src/custom-element.js]: Done

File: test/functional/test-custom-element.js (hunk 2 above)
Comment (dvoytenko) [inline test/functional/test-custom-element.js]: Will keep for now. These are older tests just refactored. I think it'd be nice to do a big sweep through all these classList.contains checks.

Comment (erwinmombay) [inline test/functional/test-custom-element.js]: sounds good. ill create an issue and will make the change.

https://github.com/ampproject/amphtml/issues/887

**Pull request merged**